### PR TITLE
Proposed take: implementation

### PR DESCRIPTION
**Pull request by hernanmd**
For ``PMMatrix`` and ``PMVector``, which answer the binomial coefficient at each element for a given integer. These are two convenience methods which makes easier to write code using contingency tables (marginals and grand totals).

### CODE REVIEW

File: src/Math-Tests-Core/PMVectorTest.class.st
Comment (hemalvarambhia) [inline src/Math-Tests-Core/PMVectorTest.class.st:173]: I'd recommend being more specific with the test name here. You could do something like

`testThatTakeYieldsACollection...`

File: src/Math-Tests-Matrix/PMMatrixTest.class.st
Comment (hemalvarambhia) [inline src/Math-Tests-Matrix/PMMatrixTest.class.st:676]: Again, here, could you be more specific about the behaviour you're testing?

**Comment by hernanmd** [comment]
HI @hemalvarambhia the selector is the one created when you right-click on a method and select "Jump to test method". This way methods are connected with their tests and recognized with the green/red circle in the method name list. It is a very common way to create test methods in Pharo.

**Comment by olekscode** [comment]
Indeed, `PMVectorTest >> testTake` is consistent with `PMVector >> take:`.
And `PMVector >> take:` is consistent with `Integer >> take:`

However, I think that `Integer >> take:` should be given a more meaningful name (e.g. `numberOfCombinationsTaken:` or `binomialCoefficientsWith:`).

Also, the implementation of `Integer >> take:` is a bit unclear:
- comment is hard to read
- names `kk`, `num`, `denom`, `factor` could be better

```Smalltalk
Integer >> take: kk
	"Return the number of combinations of (self) elements taken kk at a time.  
	For 6 take 3, this is 6*5*4 / (1*2*3).  Zero outside of Pascal's triangle.  Use a trick to go faster."
	" 6 take: 3  "

	| num denom |
	kk < 0 ifTrue: [^ 0].
	kk > self ifTrue: [^ 0].
	num := 1.
	self to: (kk max: self-kk) + 1 by: -1 do: [:factor | num := num * factor].
	denom := 1.
	1 to: (kk min: self-kk) do: [:factor | denom := denom * factor].
	^ num // denom
```

But this is outside PolyMath
We should create an issue for that in [pharo repository](https://github.com/pharo-project/pharo/)

File: src/Math-Tests-Core/PMVectorTest.class.st
Comment (olekscode) [reply src/Math-Tests-Core/PMVectorTest.class.st:173]: I think that the name is good, but this test could be split in two:
```Smalltalk
testTake
	| vector |
	vector := #(1 2 3) asPMVector.
	self assert: (vector take: 2) equals: #(0 1 3) asPMVector.
	self assert: (vector take: 3) equals: #(0 0 1) asPMVector.

testTakeEmpty
	| vector |
	vector := #() asPMVector. 
	self assert: (vector take: 2) equals: #() asPMVector.
```

Comment (hemalvarambhia) [reply src/Math-Tests-Core/PMVectorTest.class.st:173]: I'd disagree that the name is good because, looking at the code snippet, I can't tell how `vector take: 2` results in `[0, 1, 3]`

Comment (hemalvarambhia) [reply src/Math-Tests-Core/PMVectorTest.class.st:173]: Other test you can add include:
- when the argument value is more than the dimension of the vector; and
- the argument of take is 0 or negative

What happens in those cases?

**Comment by hemalvarambhia** [comment]
@hernanmd with the way you describe generating test methods in pharo, I read from that that this implies a 1:1 correspondence between message to it's test ie a method has only one test and that you must have a test method of the form `test{methodName}` . Have I understood you correctly? What happens when we write multiple tests per message? What does Pharo do then? Assuming I have understood you, whilst they may be good for navigation, we lose the ability to understand the tests as, potentially, we end up with tests that are hard to understand. In Gerard Mezsaros' 'XUnit Test Patterns' work, he recommends writing tests for humans and values tests as documentation.

As a compromise would you be will to rename the test method names to something like

`testTakeReturnsXYZWhenScenario`?

**Comment by hemalvarambhia** [comment]
Another question I have is does this message we're introducing (`take`) make sense from a domain perspective? Is this an understood concept in the vector and matrix algebra?

**Comment by hemalvarambhia** [comment]
I had a quick look at contingency tables and marginal totals on the wiki, and it looks like these ideas are from Statistics. Should we introduce this idea in a Statistics module in PolyMath?

**Comment by hemalvarambhia** [comment]
@SergeStinckwich Could you enable the check that a PR needs approval from a maintainer before the merge button goes green? We seem to have lost this.

**Comment by SergeStinckwich** [comment]
> @SergeStinckwich Could you enable the check that a PR needs approval from a maintainer before the merge button goes green? We seem to have lost this.

I think there was nothing like that. There was only a review check.

**Comment by hernanmd** [comment]
@hemalvarambhia I read your question. Nothing prevents creating additional tests for a single method. And there is no limit in creating 1,2,3,N tests. What I described is a facility from the System Browser to jump and execute the method test directly from the Browser. Other test implementations of the #take: could have more descriptive method names, but it will not be linked with the method (because they use a different selector sub-name as part of the selector),

@SergeStinckwich let me know if you are ok with continue using the test{Selector} pattern auto-generated from Calypso menu which enables linking methods with its tests. Or use the suggestion from hemalvarambhia?

File: src/Math-Core/PMVector.class.st
Comment (hemalvarambhia) [inline src/Math-Core/PMVector.class.st:311]: @hernanmd I don't recall learning about the idea described in the comment in Linear Algebra (Pure Mathematics). This could be my ignorance so could you point me to a reference?

If this isn't a Pure Mathematics idea, and given it looks like one from Statistics, should we categorize it as such? Open question

**Review by hemalvarambhia — APPROVED** [review]

**Pull request merged**
